crates/wasmtime/src/sync_nostd.rs

```diff
@@ -0,0 +1,249 @@
+//! Synchronization primitives for Wasmtime for `no_std`.
+//!
+//! These primitives are intended for use in `no_std` contexts are are not as
+//! full-featured as the `std` brethren. Namely these panic and/or return an
+//! error on contention. This serves to continue to be correct in the fact of
+//! actual multiple threads, but if a system actually has multiple threads then
+//! something will need to change in the Wasmtime crate to enable the external
+//! system to perform necessary synchronization.
+//!
+//! See a brief overview of this module in `sync_std.rs` as well.
+
+use core::cell::UnsafeCell;
+use core::mem::MaybeUninit;
+use core::ops::{Deref, DerefMut};
+use core::sync::atomic::{AtomicU32, AtomicU8, Ordering};
+
+pub struct OnceLock<T> {
+    val: UnsafeCell<MaybeUninit<T>>,
+    state: AtomicU8,
+}
+
+unsafe impl<T: Send> Send for OnceLock<T> {}
+unsafe impl<T: Sync> Sync for OnceLock<T> {}
+
+const UNINITIALIZED: u8 = 0;
+const INITIALIZING: u8 = 1;
+const INITIALIZED: u8 = 2;
+
+impl<T> OnceLock<T> {
+    pub const fn new() -> OnceLock<T> {
+        OnceLock {
+            state: AtomicU8::new(UNINITIALIZED),
+            val: UnsafeCell::new(MaybeUninit::uninit()),
+        }
+    }
+
+    pub fn get_or_init(&self, f: impl FnOnce() -> T) -> &T {
+        if let Some(ret) = self.get() {
+            return ret;
+        }
+        self.try_init::<()>(|| Ok(f())).unwrap()
+    }
+
+    pub fn get_or_try_init<E>(&self, f: impl FnOnce() -> Result<T, E>) -> Result<&T, E> {
+        if let Some(ret) = self.get() {
+            return Ok(ret);
+        }
+        self.try_init(f)
+    }
+
+    fn get(&self) -> Option<&T> {
+        match self.state.load(Ordering::Acquire) {
+            INITIALIZED => Some(unsafe { (*self.val.get()).assume_init_ref() }),
```

Comment:
This isn't going to accidentally bind `INITIALIZED` as a fresh variable, right? I always forget the details of when that can happen or not... Might be better to use an `if`/`else` here since there is just one checked-for case and one default case.

Reply:
Nah yeah we're ok to have the `match` here, I generally rely on the unused variables lint to catch mistakes like this. (I'll go ahead and switch it anyway to an `if`/`else` here)